### PR TITLE
SDK Custom Storage async setters

### LINKED ISSUE #22628: [SDK] Custom Storage configuration sometimes causes errors
### Describe the Bug

When configuring a custom storage object in the SDK, which uses async `get` and `set` functions, sometimes bug #17770 occurs due to the `storage.set` call not being awaited in the SDK. The bug occurs randomly because it depends on network stability and how fast the promises are resolved.

FYI, we followed this guide to set up the custom storage: https://docs.directus.io/guides/sdk/authentication.html#configure-custom-storage

According to the type definition, bot the `storage.set` and `storage.get` calls can be promises: https://github.com/directus/directus/blob/7e6980dfdf4a09b06455ac63ae9469dcb382a81a/sdk/src/auth/types.ts#L19

But the `storage.set` calls are not being awaited in the SDK.

See:
- unawaited `storage.set` call: https://github.com/directus/directus/blob/7e6980dfdf4a09b06455ac63ae9469dcb382a81a/sdk/src/auth/composable.ts#L72
- `resetStorage` should be async (because next line calls `storage.set`): https://github.com/directus/directus/blob/7e6980dfdf4a09b06455ac63ae9469dcb382a81a/sdk/src/auth/composable.ts#L41
- `resetStorage` call during refresh (leads to #17770 sometimes when custom storage is configured, depends on how fast the promises are resolved): https://github.com/directus/directus/blob/7e6980dfdf4a09b06455ac63ae9469dcb382a81a/sdk/src/auth/composable.ts#L90

### To Reproduce

1. Configure a custom storage implementation that is async in bot the getter and setter.
2. Now call refresh a bunch of times.
3. The bug depends on your implementation of the setter, so it's hard to reproduce. If the SDK awaits all `storage.get` and `storage.set` calls then there shouldn't be any issues.

Why does this bug happen:
- Let's say you call `refresh()` which then calls `resetStorage()`, which calls `storage.set()` (but does not await it).
- Now `resetStorage()` is gonna do its thing, but the setter is async and takes a while. (a long while due to random network instability)
- Meanwhile you got the new access token and call `storage.set()` a second time but this time the network was fine and sent it through immediately.
- Now the promise from `resetStorage()` comes back to haunt us and deletes the newly set token after the refresh function succeeded. It took a while due to network instability and ran **after** the 2nd `storage.set()` call.

### Directus Version

v10.11.0

### Hosting Strategy

Self-Hosted (Docker Image)

### Database

_No response_

### PR DESCRIPTION
**Pull request by br41nslug**
## Scope

What's changed:

- Colored a bunch of functions with async
- awaiting the storage setter in various places

## Potential Risks / Drawbacks

- Might require a change for people relying on this, altough the behavior should stay the same as now if they dont await the setter

## Review Notes / Questions

- I would like to lorem ipsum

Fixes #22628

### CODE REVIEW

File: sdk/src/auth/composable.ts [hunk 1]
```diff
@@ -87,7 +86,7 @@ export const authentication = (mode: AuthenticationMode = 'cookie', config: Part
 		const refresh = async () => {
 			const awaitRefresh = async () => {
```
Comment (paescuj) [inline sdk/src/auth/composable.ts:87]: A tad unrelated but I think we can omit the inner `awaitRefresh` wrapping here and directly return the request promise:
```ts
		const refresh = async () => {
			const authData = await storage.get();
			await resetStorage();

			const fetchOptions: RequestInit = {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
				},
			};

			if ('credentials' in authConfig) {
				fetchOptions.credentials = authConfig.credentials;
			}

			const body: Record<string, string> = { mode };

			if (mode === 'json' && authData?.refresh_token) {
				body['refresh_token'] = authData.refresh_token;
			}

			fetchOptions.body = JSON.stringify(body);

			const requestUrl = getRequestUrl(client.url, '/auth/refresh');

			return request<AuthenticationData>(requestUrl.toString(), fetchOptions, client.globals.fetch).then((data) =>
				setCredentials(data).then(() => data),
			);
		};
```

**Review by paescuj — APPROVED** [review]

**Review by rijkvanzanten — APPROVED** [review]

File: sdk/src/auth/composable.ts (hunk 1 above)
Comment (rijkvanzanten) [reply sdk/src/auth/composable.ts:87]: Let's merge the current solution first, as that solves the issue without other side effects. @br41nslug if you feel this is worth implementing as an improvement, lets do it in a separate PR 👍🏻

Comment (paescuj) [reply sdk/src/auth/composable.ts:87]: Sure 👍

Comment (br41nslug) [reply sdk/src/auth/composable.ts:87]: This approach is missing the `refreshPromise` thats used to check if the SDK is currently doing a refresh call it needs to wait for.

Comment (paescuj) [reply sdk/src/auth/composable.ts:87]: Hmm right 😅👌

**Pull request merged**